tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -3034,6 +3033,73 @@ static LogicalResult LegalizeFloatingPointPrelu(Operation* op,
   return success();
 }
 
+// Support both PReLU and leaky ReLU. The computation of PReLU is very similar
+// to leaky Relu that, having a parameter `alpha` to control a slope for
+// negative values but alpha in PReLU is a learned tensor, so compared with
+// leaky ReLU, an extra element-wise multiply alpha and input is needed for
+// PReLU.
+static LogicalResult LegalizeQuantizedLeakyAndPrelu(Operation* op,
+                                                    PatternRewriter& rewriter,
+                                                    Value input,
+                                                    double alpha_scale,
+                                                    ShapedType output_type) {
+  ShapedType rescale_type = output_type.clone(rewriter.getI32Type());
+  ShapedType input_type = input.getType().dyn_cast<ShapedType>();
+
+  UniformQuantizedType input_qtype =
+      input_type.getElementType().cast<UniformQuantizedType>();
+
+  UniformQuantizedType output_qtype =
+      output_type.getElementType().cast<UniformQuantizedType>();
+
+  double scale_alpha =
+      input_qtype.getScale() * alpha_scale / output_qtype.getScale();
+
+  double scale_identity = input_qtype.getScale() / output_qtype.getScale();
+
+  // Implement PReLU and leaky ReLU as:
+  //   rescaled_in = rescale(in)
+  //   rescaled_identity_in = rescale(in, scale_identity)
+  //   alpha_in = if (PReLU) ? mul(in, alpha) : in
+  //   rescaled_alpha_in = rescale(alpha_in, scale_alpha)
+  //   cond_result = greater_equal(rescaled_in, 0)
+  //   output = select(cond_result, rescaled_identity_in, rescaled_alpha_in)
+
+  Value op_rescale_in =
+      buildRescaleToInt32(rewriter, op, input, 1.0, input_qtype.getZeroPoint());
+
+  Value const_zero = getTosaConstTensorSingleI32(rewriter, op, 0);
+  Value op_ge = CreateOpAndInfer<tosa::GreaterEqualOp>(
+      rewriter, op->getLoc(), rescale_type.clone(rewriter.getI1Type()),
+      op_rescale_in, const_zero);
+
+  Value op_alpha_in;
+  if (auto tfl_prelu_op = dyn_cast<TFL::PReluOp>(op)) {
```

Comment:
You can cleanup the if-else block below. Set op_alpha_in to input initially. Check that op is either PreluOp or LeakyReluOp, then return a rewriter error if its not. Finally do you existing prelu op special case. It avoids multiple nested / interacting if-else statements.

Reply:
Sound good. Check the newer patch to see if the code matches your expectation, please.